generators/beamAndVertexGenerator.cc

```diff
@@ -118,80 +201,201 @@ void beamAndVertexGenerator::randomizeBeamfileStartingPosition() {
 
 
 bool beamAndVertexGenerator::check() const {
-	if(_beamTree or _simpleSimulation) {
-		return true;
-	} else {
-		return false;
+	if( not (_beamTree or _simulationMode == simpleSimulation))  return false;
+	if (_momentumPDFMean != 0.0){// do momentum resolution correction
+		if (_simulationMode != simulationFromMomentaCorrectMomentumResolution){
+			printErr << "Wrong simulation mode for for correcting the resolution in the beam momentum" << endl;
+			return false;
+		}
+		if (_momentumPDFSigma == 0.0 or _momentumResolution == 0.0){
+			printErr << "momentumResolution, momentumPDFSigma, and momentumPDFMean must be defined to use resolution correction" << endl;
+			return false;
+		}
 	}
+	return true;
 }
 
-
 bool beamAndVertexGenerator::event(const Target& target, const Beam& beam) {
+	switch(_simulationMode){
+		case simpleSimulation:
+			return eventSimple(target, beam);
+			break;
+		case simulationFromMomenta:
+			return eventFromMomenta(target, beam);
+			break;
+		case simulationFromMomentaCorrectMomentumResolution:
+			return eventFromMomentaCorrectMomentumResolution(target, beam);
+			break;
+		case simulationFromInclinations:
+			return eventFromInclinations(target, beam);
+			break;
+		default:
+			printErr << "Unknown simulation mode" << endl;
+			return false;
+			break;
+	}
+}
+bool beamAndVertexGenerator::eventSimple(const Target& target, const Beam& beam) {
 	double z = getVertexZ(target);
 	TRandom3* randomGen = randomNumberGenerator::instance()->getGenerator();
-	if(_simpleSimulation) {
-		double x;
-		double y;
-		double radius = std::sqrt(randomGen->Uniform(0, target.radius * target.radius));
-		randomGen->Circle(x, y, radius);
-		_vertex.SetXYZ(target.position.X() + x,
-		               target.position.Y() + y,
-		               z);
-		// throw magnituide of beam momentum
-		const double pBeam = randomGen->Gaus(beam.momentum, beam.momentumSigma);
-		// throw beam inclination
-		const double dxdz = randomGen->Gaus(beam.DxDz, beam.DxDzSigma);
-		const double dydz = randomGen->Gaus(beam.DyDz, beam.DyDzSigma);
-		// construct tilted beam momentum Lorentz vector
-		const double pz        = pBeam / sqrt(1 + dxdz * dxdz + dydz * dydz);
-		const double px        = dxdz * pz;
-		const double py        = dydz * pz;
-		const double EBeam     = sqrt(pBeam * pBeam + beam.particle.mass2());
-		_beam.SetXYZT(px, py, pz, EBeam);
+	double x;
+	double y;
+	double radius = std::sqrt(randomGen->Uniform(0, target.radius * target.radius));
```

Comment:
Why is the radius picked randomly? TRandom::Circle() generates random x and y within given radius.

Reply:
See ROOT documentation: https://root.cern/doc/master/classTRandom.html#a837dc7758480bd5e0e170687560bb4f8

>  Generates random vectors, uniformly distributed over a circle of given radius.
> Input : r = circle radius Output: x,y a random 2-d vector of length r 

So I think picking the radius randomly is ok, right?